mne/beamformer/_ap.py

```diff
@@ -0,0 +1,743 @@
+                                 a_tmp.transpose()])
+            perpend_spc = np.eye(act_spc.shape[0]) - act_spc
+            for dip in range(attr_dict['ndipoles']):
+                sol_tuple = _solve_active_gain_eig(dip, gain, data_cov,
+                                                   eig, perpend_spc)
+                ap_val2[dip] = np.max([x.real for x in sol_tuple[0]])
+
+            sq_idx = np.argmax(ap_val2)
+            s_ap_2[src] = sq_idx
+            sol_tuple = _solve_active_gain_eig(sq_idx, gain, data_cov,
+                                               eig, perpend_spc)
+            oris[src] = sol_tuple[1][:, [np.argmax([x.real for x
+                                                    in sol_tuple[0]])]][:, 0]
+            sub_g_proj[:, src] = np.dot(sol_tuple[2], oris[src])
+
+        logger.info('current s_ap_2 = {}'.format(s_ap_2))
+        if (itr > 0) & (s_ap_2_prev == s_ap_2):
+            # No improvement vs. previous iteration
+            logger.info('Done (optimally)')
+            break
+        if itr == attr_dict['max_iter']:
+            logger.info('Done (max iteration)')
+
+    return s_ap_2, oris, sub_g_proj
+
+
+def _calculate_free_alternating_projections(data_arr, gain,
+                                            nsources, max_iter):
+    """Calculate free-orientation alternating projection.
+
+    Parameters
+    ----------
+    data_arr : array, shape (nchannels, times)
+        Filttered evoked data array.
+    gain : array, shape (nchannels, ndipoles)
+        Gain array.
+    nsources : int
+        The number of dipoles to estimate.
+    max_iter : int
+        Maximal iteration number of AP.
+
+    Returns
+    -------
+    ap_temp_tuple : tuple
+        See: _free_phase2.
+
+    """
+    print('calculating free-orientation alternating projection')
+    ndipoles = int(gain.shape[1] / 3)
+    attr_dict = {
+        'ndipoles': ndipoles,
+        'nsources': nsources,
+        'max_iter': max_iter
+    }
+
+    data_cov = _produce_data_cov(data_arr, attr_dict)
+    # ######################################
+    # 1st Phase
+    # (a) Initialization: search the 1st source location over the entire
+    # dipoles topographies space
+    # ######################################
+
+    print(' 1st phase : ')
+    ap_temp_tuple = _free_phase1a(attr_dict, gain, data_cov)
+    # ap_temp_tuple = (s_ap, oris, sub_g_proj)
+
+    # ######################################
+    # (b) Now, add one source at a time
+    # ######################################
+
+    ap_temp_tuple = _free_phase1b(attr_dict, gain, data_cov, ap_temp_tuple)
+    # ap_temp_tuple = (s_ap, oris, sub_g_proj)
+    print('current s_ap = {}'.format(ap_temp_tuple[0]))
+
+    # #####################################
+    # 2nd phase
+    # #####################################
+
+    ap_temp_tuple = _free_phase2(ap_temp_tuple, attr_dict, data_cov, gain)
+
+    return ap_temp_tuple
+
+
+def free_ori_ap(wh_data, gain, nsources, forward, max_iter):
```

Comment:
```suggestion
def _free_ori_ap(wh_data, gain, nsources, forward, max_iter):
```
this should be private

Reply:
done